File: export-templates/javascript/src/components/tools.js
const { DynamicStructuredTool } = require('@langchain/core/tools');
const { AIMessage, HumanMessage, SystemMessage } = require('@langchain/core/messages');
const { ChatOpenAI } = require('@langchain/openai');
const { getRelevantExamples } = require('./examples');
const { createAdhocLogFile } = require('./logger');
const { extractFunctionCode } = require('./parser');
const { ADHOC_PROMPT_TEXT, LIBS, FUNCTIONS, IS_LAMBDA, TASK_TOOL_CALL_DESCRIPTION_TEMPLATES } = require('../constants');
const vm = require('vm');

const TOOL_TIMEOUT = parseInt(process.env.TOOL_TIMEOUT || '60000');


function captureAndProcessOutput(func, args = [], streamWriter) {
  return new Promise((resolve, reject) => {
    let resolved = false;

    const customLog = (arg) => {
      if (resolved) return;
      resolved = true;

      let result = arg;
      if (typeof result === 'string') {
        try {
          result = JSON.parse(result);
        } catch {
          // leave as string
        }
      }
      resolve(result);
    };

    // Signature-preserving safe wrapper
    function makeSafeWrapper(fn) {
      const safe = function (...args) {
        try {
          return fn.apply(null, args);
        } catch (e) {
          return { error: e.message };
        }
      };
      Object.defineProperty(safe, 'name', { value: fn.name, configurable: true });
      return Object.freeze(safe);
    }

    // Recursively wrap all functions in an object
    function wrapObject(obj) {
      if (typeof obj !== 'object' || obj === null) return obj;

      const wrapped = {};
      for (const [key, value] of Object.entries(obj)) {
        if (typeof value === 'function') {
          wrapped[key] = makeSafeWrapper(value);
        } else if (typeof value === 'object' && value !== null) {
          wrapped[key] = Object.freeze(wrapObject(value)); // nested object
        } else {
          wrapped[key] = value;
        }
      }
      return Object.freeze(wrapped);
    }

    const frozenContext = {
      ...wrapObject({ streamWriter }),
      ...wrapObject(LIBS),
      ...wrapObject(FUNCTIONS),
      console: Object.freeze({
        log: customLog,
        warn: console.warn,
        error: console.error,
      }),
    };

    const context = vm.createContext(frozenContext);

    const funcString = typeof func === 'function' ? func.toString() : func;

    const wrappedCode = `
      (async () => {
        const userFunc = ${funcString};
        const result = await userFunc(...${JSON.stringify(args)});
        console.log(result);
      })().catch(console.error);
    `;

    try {
      const script = new vm.Script(wrappedCode, { timeout: TOOL_TIMEOUT });
      script.runInContext(context);
    } catch (err) {
      reject(err);
    }

    setTimeout(() => {
      if (!resolved) {
        resolved = true;
        reject(new Error(`Function execution timed out after ${TOOL_TIMEOUT} ms`));
      }
    }, TOOL_TIMEOUT);
  });
}


// function captureAndProcessOutput(func, args = [], streamWriter) {
//   return new Promise((resolve, reject) => {
//     const customLog = (arg) => {
//       // Assuming we only need the first argument as tasks return a single object
//       let result = arg;

//       if (typeof result === 'string') {
//         try {
//           result = JSON.parse(result);
//         } catch (error) {
//           // If parsing fails, keep it as a string
//           // No need to do anything here as result is already a string
//         }
//       } else {
//         // Convert non-string types to string
//         result = String(result);
//       }
      
//       resolve(result);
//     };

//     // Create a context object with all the necessary functions and variables
//     const context = {
//       require,
//       console: { log: customLog, warn: console.warn, error: console.error  },
//       // Add all the functions and variables from the local scope that the function might need
//       streamWriter: streamWriter,
//       ...LIBS,
//       ...FUNCTIONS
//     };

//     // Convert the function to a string
//     const funcString = typeof func === 'function' ? func.toString() : func;

//     // Create a new function with the context as its scope
//     const contextFunction = new Function(...Object.keys(context), `
//       return async function() {
//         try {
//           global.streamWriter = streamWriter;

//           return await (${funcString}).apply(this, arguments);
//         } catch (error) {
//           console.warn("Error executing tool:", error);
//           throw error;
//         }
//       }
//     `)(...Object.values(context));

//     try {
//       // Execute the function with the provided arguments
//       contextFunction(...args).then(resolve).catch(reject);

//       setTimeout(() => {
//         reject(new Error(`Function execution timed out after ${TOOL_TIMEOUT} ms`));
//       }, TOOL_TIMEOUT);
//     } catch (error) {
//       reject(error);
//     }
//   });
// }

// Capture stdout of tasks
async function toolWrapper(func, args, streamWriter) {
  // todo: make sure the args are in the correct positional order
  // this code extracts the args map from the object and passes them as individual value arguments
  try {
    const result = await captureAndProcessOutput(func, Object.values(args), streamWriter);
    // Ensure the result is a string
    return (typeof result === 'object' ? JSON.stringify(result) : String(result));
  } catch (error) {
    console.error(`Error in toolWrapper: ${error.message}`);
    return `Error: ${error.message}`;
  }
}

function createToolsFromSchemas(schemas) {
  const tools = [];
  for (const schema of schemas) {
    let description = schema.description;
    if (schema.returns_description) {
      description += ` Returns: ${schema.returns_description}`;
    }

    const tool = new DynamicStructuredTool({
      name: schema.name,
      description,
      schema: schema.schema,
      func: schema.name === 'run_adhoc_task' ? schema.func : async (args, streamWriter) => await toolWrapper(schema.func, args, streamWriter)
    });

    tools.push(tool);
  }
  return tools;
}

const apiKey = process.env.OPENAI_API_KEY;
const model = process.env.OPENAI_MODEL;

const adhocLLM = new ChatOpenAI({
  apiKey,
  model,
  configuration: {
    defaultHeaders: {
      'Connection': 'keep-alive',
      'Keep-Alive': 'timeout=900'
    },
  },
});

async function generateAndExecuteAdhoc(userInput, streamWriter, maxRetries = 5) {
  let retryCount = 0;
  const errors = [];
  let previousCode = null;

  // Get relevant examples
  const relevantExamples = await getRelevantExamples(userInput);

  while (retryCount < maxRetries) {
    try {
      // Prepare the prompt with error information if available
      let errorContext = "";
      if (errors.length > 0) {
        errorContext = `This is retry attempt ${retryCount}.\nPrevious errors:\n`;
        errors.forEach((error, index) => {
          const modifiedError = error.includes("The request cannot be fulfilled using the available functions")
            ? "Syntax error" // faking a syntax error seems to improve the retry success rate
            : error;
          errorContext += `${index + 1}. ${'-'.repeat(40)}\n${modifiedError}\n\n`;
        });
        
        if (previousCode) {
          errorContext += `Previous code:\n\`\`\`javascript\n${previousCode}\n\`\`\`\n\n`;
        }
        errorContext += `The previously generated code failed because of these issues, please re-write the code to address them.\nIf the errors are not clear or useful please write the code again based on the instructions and available functions.\nAssume you are more capable than the agent that generated the previous attempt and you can make better decisions.`;
      }

      const exampleMessages = relevantExamples.flatMap((example) => [
        new HumanMessage(example.task),
        new AIMessage(example.code)
      ]);

      // Use the generic language model for the completion
      const messages = [
        new SystemMessage("You are a useful technical assistant."),
        new SystemMessage(ADHOC_PROMPT_TEXT),
        ...exampleMessages,
        new HumanMessage(`${userInput}\n\n${errorContext}`)
      ];

      const response = await adhocLLM.invoke(messages);

      if (response.content.includes('The request cannot be fulfilled using the available functions')) {
        throw new Error(response.content);
      }

      const functionCode = extractFunctionCode(response.content);

      if (!functionCode) {
        throw new Error(`Failed to parse function code: ${response.content}`);
      }

      previousCode = functionCode;

      console.log("Generated code:", functionCode);

      const result = await captureAndProcessOutput(functionCode, [], streamWriter);
      
      if (!IS_LAMBDA) createAdhocLogFile(userInput, functionCode, result);
      
      return result;
    } catch (e) {
      const errorMessage = e.message;
      console.error(`Error during execution (attempt ${retryCount + 1}): ${errorMessage}`);
      errors.push(errorMessage);
      retryCount++;

      if (retryCount === maxRetries) {
        console.error(`Max retries (${maxRetries}) reached. Aborting.`);
        if (!IS_LAMBDA) createAdhocLogFile(userInput, previousCode, '', new Error(`Max retries reached. Last error: ${errorMessage}`));
        throw new Error(`Max retries reached. Last error: ${errorMessage}`);
      }

      await new Promise(resolve => setTimeout(resolve, 500));

      console.log(`Retrying... (attempt ${retryCount} of ${maxRetries})`);
    }
  }

  // This line should never be reached, but just in case
  throw new Error("Unexpected error occurred");
}

function getToolCallDescription(toolName, args) {
  const toolCallDescriptionTemplate = TASK_TOOL_CALL_DESCRIPTION_TEMPLATES[toolName];
  if (!toolCallDescriptionTemplate) {
    return null;
  }

  return toolCallDescriptionTemplate.replace(/#(\w+)#/g, (match, p1) => args[p1]);
}

module.exports = {
  captureAndProcessOutput,
  createToolsFromSchemas,
  generateAndExecuteAdhoc,
  getToolCallDescription
};

File: export-templates/python/src/components/tools.py
import os
import asyncio
import io
import json
import sys
import traceback
from typing import Dict, Any, List
from langchain_core.tools import StructuredTool
from langchain_openai import ChatOpenAI
from langchain.schema import SystemMessage, HumanMessage, AIMessage
from langchain.chat_models.base import BaseChatModel
from functools import partial
from contextlib import redirect_stdout
from components.examples import get_relevant_examples
from components.parser import extract_function_code
from components.logger import create_adhoc_log_file
from constants import ADHOC_PROMPT_TEXT, LIBS, FUNCTIONS

TOOL_TIMEOUT = int(os.getenv('TOOL_TIMEOUT', 60000)) / 1000

# todo: do we need to handle warn and error logs?
async def capture_and_process_output(func, *args, emit_event=None, **kwargs):
    f = io.StringIO()
    try:
        async def execute():
            with redirect_stdout(f):
                builtins_dict = sys.modules['builtins'].__dict__
                
                def info(*args, **kwargs):
                    if emit_event:
                        emit_event(*args, **kwargs)
                
                builtins_dict['info'] = info

                if asyncio.iscoroutinefunction(func):
                    return await func(*args, **kwargs)
                else:
                    return func(*args, **kwargs)

        result = await asyncio.wait_for(execute(), timeout=TOOL_TIMEOUT)

        output = f.getvalue()
        
        try:
            processed_result = json.loads(output)
        except json.JSONDecodeError:
            processed_result = output.strip()
        
        return processed_result
    except asyncio.TimeoutError:
        print(f"Execution timed out after {TOOL_TIMEOUT} seconds", file=sys.stderr)
        raise
    except Exception as e:
        print(f"Error executing tool:", file=sys.stderr)
        traceback.print_exc()
        raise

# Capture stdout of tasks
async def tool_wrapper(func, *args, emit_event=None, **kwargs):
    # todo: make sure the args are in the correct positional order
    # this code extracts the args map from the object and passes them as individual value arguments
    arguments = args[0] if isinstance(args, tuple) and len(args) == 1 else args
    arguments_json = json.dumps(arguments)
    arguments_map = json.loads(arguments_json)
    positional_args = list(arguments_map.values())
            
    return await capture_and_process_output(func, *positional_args, emit_event=emit_event, **kwargs)

def create_tools_from_schemas(schemas: Dict[str, Dict[str, Any]]) -> List[StructuredTool]:
    tools = []
    for name, schema in schemas.items():
        description = schema["description"]
        if "returns_description" in schema:
            description += f" Returns: {schema['returns_description']}"
        
        if name == "run_adhoc_task":
            func = schema["function"]
        else:
            func = partial(tool_wrapper, schema["function"])
        
        tool = StructuredTool(
            name=name,
            description=description,
            args_schema=schema["args_schema"],
            coroutine=func,
            metadata={"output": schema["output"]}
        )
        tools.append(tool)
    return tools

api_key = os.getenv('OPENAI_API_KEY')
model = os.getenv('OPENAI_MODEL')

adhoc_llm: BaseChatModel = ChatOpenAI(api_key=api_key, model=model)

async def execute_generated_function(function_code):
    # Create a temporary module to execute the function
    import types
    module = types.ModuleType("temp_module")
    
    # Add necessary imports to the module
    module.__dict__.update({
        **{func.__name__: func for func in LIBS},
        **{func.__name__: func for func in FUNCTIONS},
        'json': json,
    })
    
    # Execute the function code in the module's context
    exec(function_code, module.__dict__)
    
    # Call the doTask function
    result = module.doTask()
    
    # If the result is a coroutine, await it
    if asyncio.iscoroutine(result):
        result = await result
    
    return result

# Adhoc task execution
async def generate_and_execute_adhoc(user_input: str, max_retries: int = 5):
    retry_count = 0
    errors = []
    previous_code = None

    # Get relevant examples
    relevant_examples = get_relevant_examples(user_input)

    while retry_count < max_retries:
        try:
            # Prepare the prompt with error information if available
            error_context = ""
            if errors:
                error_context = f"This is retry attempt ${retry_count}.\nPrevious errors:\n"
                for index, error in enumerate(errors, 1):
                    # faking a syntax error seems to improve the retry success rate
                    modified_error = "Syntax error" if "The request cannot be fulfilled using the available functions" in error else error
                    error_context += f"{index}. {'-' * 40}\n{modified_error}\n\n"
                
                if previous_code:
                    error_context += f"Previous code:\n```python\n{previous_code}\n```\n\n"
                
                error_context += "The previously generated code failed because of these issues, please re-write the code to address them.\nIf the errors are not clear or useful please write the code again based on the instructions and available functions.\nAssume you are more capable than the agent that generated the previous attempt and you can make better decisions."
            
            example_messages = [
                {"role": "human" if i % 2 == 0 else "assistant", "content": content}
                for example in relevant_examples
                for i, content in enumerate([example["task"], example["code"]])
            ]
            
            # Use the generic language model for the completion
            messages = [
                SystemMessage("You are a useful technical assistant."),
                SystemMessage(ADHOC_PROMPT_TEXT),
                *[HumanMessage(content=msg["content"]) if msg["role"] == "human" else AIMessage(content=msg["content"]) for msg in example_messages],
                HumanMessage(content=f"{user_input}\n\n{error_context}")
            ]
            response = await adhoc_llm.agenerate([messages])

            if 'The request cannot be fulfilled using the available functions' in response.generations[0][0].text:
                raise ValueError(response.generations[0][0].text)
            
            function_code = extract_function_code(response.generations[0][0].text)

            if not function_code:
                raise ValueError(f"Failed to parse function code: {response.generations[0][0].text}")
            
            previous_code = function_code

            print("Generated code:", function_code, flush=True)

            result = await capture_and_process_output(execute_generated_function, function_code)
            
            create_adhoc_log_file(user_input, function_code, result)
            
            return result
        except Exception as e:
            error_message = str(e)
            print(f"Error during execution (attempt {retry_count + 1}): {error_message}", flush=True)
            errors.append(error_message)
            retry_count += 1

            if retry_count == max_retries:
                print(f"Max retries ({max_retries}) reached. Aborting.", flush=True)
                create_adhoc_log_file(user_input, previous_code, '', error=f"Max retries reached. Last error: {error_message}")
                raise ValueError(f"Max retries reached. Last error: {error_message}")

            print(f"Retrying... (attempt {retry_count} of {max_retries})", flush=True)

    # This line should never be reached, but just in case
    raise ValueError("Unexpected error occurred")